### PR TITLE
PHPS: made valid() stricter to avoid overflows

### PR DESCRIPTION
**Pull request by AlekseyCherepanov**
Related to #4971.

### CODE REVIEW

**Comment by AlekseyCherepanov** [comment]
One self-test was commented out with such comment: `fails $HEX$ IS required by dyna`. I don't know if it was true then, but now it is not true. After my fix it can be loaded well. So I uncommented the self-test. Also I added one.

Somehow dynamic's valid is quite permissive. There is sample that can be loaded:
```
$dynamic_6$612f95cf86c4e196f6b627bab3f01b62$HEX$$$$$$$
```

The sample does not cause a crash. So I am going to open a separate issue.

**Comment by AlekseyCherepanov** [comment]
`--test` and `--fuzz` are ok.

File: src/PHPS_fmt_plug.c [hunk 1]
```diff
@@ -162,10 +163,12 @@ static int phps_valid(char *ciphertext, struct fmt_main *self)
 
 		if (!val)
 			return 0;
-		cp = strrchr(ciphertext, '$');
-		if (strstr(ciphertext, "$HEX$"))
-			wanted_len += 3; // salt is in hex.
-		return cp && strlen(cp) == wanted_len;
+		cp = &ciphertext[11 + MD5_HEX_SIZE];
```
Comment (solardiz) [inline src/PHPS_fmt_plug.c:166]: How do we know(?) we're not(?) over-jumping beyond `ciphertext` length here?

Comment (AlekseyCherepanov) [reply src/PHPS_fmt_plug.c:166]: I hoped that `pDynamic_6->methods.valid` would catch shorter ciphertexts. I tried: it seems to reject them as expected.

**Review by solardiz — APPROVED** [review]

**Pull request merged**